multiworld/envs/real_sawyer/sawyer_reaching.py

```diff
@@ -0,0 +1,683 @@
+        return reward
+
+
+    def _randomize_desired_end_effector_pose(self):
+        #self.desired = np.random.uniform(self.safety_box_lows, self.safety_box_highs, size=(1, 3))[0]
+        assert False # Don't randomize goal inside reset! In MultitaskEnv this is done by set_goal
+        high = self.pd_safety_box_high
+        low = self.pd_safety_box_low
+        dx = np.random.uniform(low[0], high[0])
+        dy = np.random.uniform(low[1], high[1])
+        dz = np.random.uniform(low[2], high[2])
+        self.desired =  np.array([dx, dy, dz])
+
+    def compute_her_reward_np(self, observation, action, next_observation):
+        return -np.linalg.norm(next_observation['endpoint_pose'][:3] - next_observation['goal_state'])
+
+    def reset(self):
+        self.in_reset = True
+        self._safe_move_to_neutral()
+        self.in_reset = False
+        return self._get_observation()
+
+
+    def step(self, action):
+        self._act(action)
+        observation = self._get_observation()
+        reward = self.reward() * self.reward_magnitude
+        done = False
+        info = {}
+        info['goal'] = self.desired.copy()
+        return observation, reward, done, info
+
+    def _get_statistics_from_paths(self, paths):
+        statistics = OrderedDict()
+        stat_prefix = 'Test'
+        distances_from_target, final_position_distances = self._extract_experiment_info(paths)
+        statistics.update(self._update_statistics_with_observation(
+            distances_from_target,
+            stat_prefix,
+            'End Effector Distance from Target'
+        ))
+        statistics.update(self._update_statistics_with_observation(
+            final_position_distances,
+            stat_prefix,
+            'Final End Effector Distance from Target'
+        ))
+        return statistics
+
+    def _extract_experiment_info(self, paths):
+        obsSets = [path["next_observations"] for path in paths]
+        goalSet = [path['goals'][0] for path in paths]
+        positions = []
+        desired_positions = []
+        distances = []
+        final_positions = []
+        final_desired_positions = []
+        for obsSet, goal in zip(obsSets, goalSet):
+            for observation in obsSet:
+                pos = np.array(observation[21:24])
+                distances.append(np.linalg.norm(pos - goal))
+                positions.append(pos)
+                desired_positions.append(goal)
+            final_positions.append(positions[-1])
+            final_desired_positions.append(desired_positions[-1])
+
+        distances = np.array(distances)
+        final_positions = np.array(final_positions)
+        final_desired_positions = np.array(final_desired_positions)
+        final_position_distances = np.linalg.norm(final_positions - final_desired_positions, axis=1)
+        return [distances, final_position_distances]
+
+
+
+
+class SawyerXYZReachingImgMultitaskEnv(SawyerEnv, MultitaskEnv):
```

Comment:
We'll have the ImageEnv wrapper for images vs no images. So I believe this class and the one below are unnecessary. I think we want to have a separate class for pushing vs reaching, because they may have different safety boxes, different reset procedures, etc. So there should probably be a SawyerXYZEnv that would be used if you just wanted to do reaching. Then a class SawyerXYZPushingEnv which is a child of SawyerXYZEnv and contains only a little bit of code to specify safety box, reset procedure, goal setting, etc.